### PR TITLE
Add a tool for generating package dependency report

### PR DESCRIPTION
**Pull request by aws-tianquaw**
Add a tool for generating package dependency report for newly introduced packages in an image. It will provide a full list of dependencies and required versions for new marquee packages in the image release.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by aws-tianquaw** [comment]
Here is a sample report:

# Python Package Dependency Report


### Target Image Version: 1.9.0 | Base Image Version: 1.8.0

## Image Type: (GPU)
Package | Version in the Target Image | Dependencies
---|---|---
langchain-aws|0.1.6|['boto3 >=1.34.51,<1.35.0', 'langchain-core >=0.1.45,<0.3.0', 'numpy >=1.0.0,<2.0.0', 'python >=3.8.1,<4.0']
amazon-sagemaker-jupyter-ai-q-developer|1.0.3|['aws-embedded-metrics', 'boto3', 'botocore', 'jupyter-ai >=2.17.0,<3', 'jupyter-ai-magics >=2.17.0,<3', 'jupyter_server >=2.0.1,<3', 'jupyterlab >=4.0,<5', 'langchain-core', 'python >=3.8', 'sagemaker-jupyterlab-extension-common >=0.1.18,<1']
amazon-q-developer-jupyterlab-ext|3.1.0|['aiobotocore', 'boto3', 'botocore', 'jupyterlab >=4.0,<5.0', 'python >=3.6']
tensorflow-gpu|2.14.0|['tensorflow 2.14.0 cuda118py310h148f8e3_0']

## Image Type: (CPU)
Package | Version in the Target Image | Dependencies
---|---|---
langchain-aws|0.1.6|['boto3 >=1.34.51,<1.35.0', 'langchain-core >=0.1.45,<0.3.0', 'numpy >=1.0.0,<2.0.0', 'python >=3.8.1,<4.0']
amazon-sagemaker-jupyter-ai-q-developer|1.0.3|['aws-embedded-metrics', 'boto3', 'botocore', 'jupyter-ai >=2.17.0,<3', 'jupyter-ai-magics >=2.17.0,<3', 'jupyter_server >=2.0.1,<3', 'jupyterlab >=4.0,<5', 'langchain-core', 'python >=3.8', 'sagemaker-jupyterlab-extension-common >=0.1.18,<1']
amazon-q-developer-jupyterlab-ext|3.1.0|['aiobotocore', 'boto3', 'botocore', 'jupyterlab >=4.0,<5.0', 'python >=3.6']

**Review by TRNWWZ — APPROVED** [review]

**Comment by TRNWWZ** [comment]
Will this report be attached to PR automatically when using Github Action to generate PR?

**Comment by aws-tianquaw** [comment]
> Will this report be attached to PR automatically when using Github Action to generate PR?

Once the PR is merged, I'll update our CodeBuild script to automatically generate report and attach to release PRs.

File: src/package_report.py [hunk 1]
```diff
@@ -221,6 +221,35 @@ def _generate_python_package_size_report_per_image(
     return validate_result
 
 
+def _generate_python_package_dependency_report(image_config, base_version_dir, target_version_dir):
+    env_in_file_name = image_config["build_args"]["ENV_IN_FILENAME"]
+
+    base_match_spec_out = get_match_specs(base_version_dir + "/" + env_in_file_name) if base_version_dir else dict()
```
Comment (balajisankar15) [inline src/package_report.py:227]: rename to base_match_spec_in ?

File: src/package_report.py [hunk 2]
```diff
@@ -221,6 +221,35 @@ def _generate_python_package_size_report_per_image(
...
+    target_match_spec_out = get_match_specs(target_version_dir + "/" + env_in_file_name)
+
+    base_packages_match_spec_out = {k: v for k, v in base_match_spec_out.items()}
+    target_packages_match_spec_out = {k: v for k, v in target_match_spec_out.items()}
+
+    results = dict()
+    for package, target_match_spec_out in target_packages_match_spec_out.items():
+        # We only care about the newly introduced marquee packages
+        if package not in base_packages_match_spec_out:
+            # Pull package metadata from conda-forge and dump into json file
```
Comment (balajisankar15) [inline src/package_report.py:237]: For this package, you should actually query env.out to get the corresponding Match spec (which includes the version that is used during the build.)

Then we should use the env.out match spec in the search query

**Review by balajisankar15 — APPROVED** [review]

**Pull request merged**